{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "toc": true
   },
   "source": [
    "<h1>Table of Contents<span class=\"tocSkip\"></span></h1>\n",
    "<div class=\"toc\"><ul class=\"toc-item\"><li><span><a href=\"#Finding-genome-coordinates-for-MaxQuant-Derived-Peptides\" data-toc-modified-id=\"Finding-genome-coordinates-for-MaxQuant-Derived-Peptides-1\"><span class=\"toc-item-num\">1&nbsp;&nbsp;</span>Finding genome coordinates for MaxQuant Derived Peptides</a></span><ul class=\"toc-item\"><li><span><a href=\"#Retreiving-Indexes-and-Information-from-a-CSV-file\" data-toc-modified-id=\"Retreiving-Indexes-and-Information-from-a-CSV-file-1.1\"><span class=\"toc-item-num\">1.1&nbsp;&nbsp;</span>Retreiving Indexes and Information from a CSV file</a></span></li><li><span><a href=\"#Preparing-relevant-data-for-running-a-command-line-tBLASTn-search-against-the-Tb927-genome\" data-toc-modified-id=\"Preparing-relevant-data-for-running-a-command-line-tBLASTn-search-against-the-Tb927-genome-1.2\"><span class=\"toc-item-num\">1.2&nbsp;&nbsp;</span>Preparing relevant data for running a command line tBLASTn search against the Tb927 genome</a></span></li><li><span><a href=\"#Command-Line-tBLASTn-code\" data-toc-modified-id=\"Command-Line-tBLASTn-code-1.3\"><span class=\"toc-item-num\">1.3&nbsp;&nbsp;</span>Command Line tBLASTn code</a></span></li><li><span><a href=\"#Handling-BLAST-output\" data-toc-modified-id=\"Handling-BLAST-output-1.4\"><span class=\"toc-item-num\">1.4&nbsp;&nbsp;</span>Handling BLAST output</a></span></li><li><span><a href=\"#Comparison-of-orginal-(MaxQuant)-and-new-(tBLASTn)-results\" data-toc-modified-id=\"Comparison-of-orginal-(MaxQuant)-and-new-(tBLASTn)-results-1.5\"><span class=\"toc-item-num\">1.5&nbsp;&nbsp;</span>Comparison of orginal (MaxQuant) and new (tBLASTn) results</a></span></li></ul></li><li><span><a href=\"#Reveiwing-Mass-Spec-data-(MaxQuant)\" data-toc-modified-id=\"Reveiwing-Mass-Spec-data-(MaxQuant)-2\"><span class=\"toc-item-num\">2&nbsp;&nbsp;</span>Reveiwing Mass Spec data (MaxQuant)</a></span><ul class=\"toc-item\"><li><span><a href=\"#Retreiving-relevant-Information-for-new-protein-coding-genes\" data-toc-modified-id=\"Retreiving-relevant-Information-for-new-protein-coding-genes-2.1\"><span class=\"toc-item-num\">2.1&nbsp;&nbsp;</span>Retreiving relevant Information for new protein coding genes</a></span></li></ul></li></ul></div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import Bio as bio\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Finding genome coordinates for MaxQuant Derived Peptides\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "MaxQuant was used to predict new peptides and map them to the T brucei 927 (Tb927) genome available at TriTrypDB, with the intention of using these genomic coordinates to discover new protein coding gene regions. We wished to verify the genomic coordinates output from MaxQuant by re-mapping the peptides onto a newer version of the Tb927 genome (version 48) using a cistom tBLASTn set-up"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Retreiving Indexes and Information from a CSV file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Find column numbers for peptide reference (pep_ref), gene name, and ID number and define function to retrieve information on a requested feature:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "my_file = open('predicted_new_pep.csv', 'r')\n",
    "\n",
    "def get_index(infile, feature):\n",
    "    headers = (infile.readline()).split(',')\n",
    "    index = headers.index(feature)\n",
    "    infile.seek(0)\n",
    "    return index\n",
    "\n",
    "pep_ref_index = get_index(my_file, 'pep_ref')\n",
    "name_index = get_index(my_file, 'name')\n",
    "ID_index = get_index(my_file, '')\n",
    "\n",
    "my_file.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "my_file = open('predicted_new_pep.csv', 'r')\n",
    "\n",
    "def get_info(infile, index, feature):\n",
    "    feature = []\n",
    "    for line_number, line in enumerate(infile):\n",
    "        if line_number >0:\n",
    "            feature.append((line.split(','))[index])\n",
    "    infile.seek(0)\n",
    "    return feature\n",
    "\n",
    "my_file.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For each row in predicted_new_pep.csv file; read pep_ref and get just the peptide sequence, then store as a dictionary key (ensures unique peptides only) with corresponding value as an appropriate identifying fasta title (>pep{ID number}_{gene name}) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "895\n",
      "703\n"
     ]
    }
   ],
   "source": [
    "my_file = open('predicted_new_pep.csv', 'r')\n",
    "\n",
    "all_peptides = []\n",
    "pep_dict = {}\n",
    "\n",
    "\n",
    "for line_number, line in enumerate(my_file):\n",
    "    if line_number >0:                              #skip over column headers\n",
    "        listed = line.split(',')                    #split line into list of values for each column entry\n",
    "        pep_details = (listed[pep_ref_index]).split('-')  #splits up pep_ref entry to allow only peptide sequence to be selected\n",
    "        peptide = pep_details[5]                    #select just sequence\n",
    "        gene_name = (listed[name_index])            #retrieve gene name... \n",
    "        num_ID = (listed[ID_index])                 #...and ID number for identifying once in fasta format\n",
    "        fasta_title = '>pep{}_{}'.format(num_ID,gene_name) #create header for fasta with ID number and name\n",
    "        all_peptides.append(peptide)                #store all peptides as a list, in order to have for future reference\n",
    "        pep_dict[peptide] = fasta_title             #dict_key is peptide sequence, dict_value is fasta header\n",
    "\n",
    "print(len(all_peptides))  #double check to ensure non unique peptides have been removed\n",
    "print(len(pep_dict))\n",
    "\n",
    "\n",
    "my_file.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Preparing relevant data for running a command line tBLASTn search against the Tb927 genome"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Write the contents of the unique peptides dictionary out into a file in fasta format (>title \\n peptide sequence)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "fasta_out = open('pep_ref_all.fa', 'w')\n",
    "\n",
    "for entry in pep_dict.keys():\n",
    "    fasta_out.write(pep_dict[entry]+'\\n')  #write fasta header\n",
    "    fasta_out.write(entry+'\\n')            #write peptide sequence\n",
    "\n",
    "fasta_out.close()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Split fasta sequences into long (>20aa) and short (<20aa) sequences and save them in different files so as to run tBLASTn with different parameters for short/long peptides. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "peps_long = {}\n",
    "peps_short = {}\n",
    "\n",
    "fasta_out_longpeps = open('pep_ref_long.fa', 'w')\n",
    "fasta_out_shortpeps = open('pep_ref_short.fa', 'w')\n",
    "\n",
    "for entry in pep_dict.keys():\n",
    "    if len(entry)>=20:\n",
    "        fasta_out_longpeps.write(pep_dict[entry]+'\\n')  #write fasta header\n",
    "        fasta_out_longpeps.write(entry+'\\n')            #write peptide sequence\n",
    "    if len(entry)<20:\n",
    "        fasta_out_shortpeps.write(pep_dict[entry]+'\\n')  #write fasta header\n",
    "        fasta_out_shortpeps.write(entry+'\\n')            #write peptide sequence\n",
    "        \n",
    "fasta_out_longpeps.close()\n",
    "fasta_out_shortpeps.close()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Command Line tBLASTn code\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The tBLASTn set-up itself was run scripted with bash. The database 927_genome was assembled from the whole genome fasta file downloaded from TriTrypDB using the script below. Both this raw file and the resulting databse files (927_genome.nhr, 927_genome.nin and 927_genome.nsqr) are available in the associated Git repository"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "Building a new DB, current time: 12/14/2020 15:23:53\n",
      "New DB name:   /cluster/home/emarriott/927_genome\n",
      "New DB title:  TriTrypDB-48_TbruceiTREU927_Genome.fasta\n",
      "Sequence type: Nucleotide\n",
      "Deleted existing Nucleotide BLAST database named /cluster/home/emarriott/927_genome\n",
      "Keep MBits: T\n",
      "Maximum file size: 1000000000B\n",
      "Adding sequences from FASTA; added 131 sequences in 0.89407 seconds.\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "#!/bin/bash\n",
    "\n",
    "#$ -cwd\n",
    "\n",
    "makeblastdb -in TriTrypDB-48_TbruceiTREU927_Genome.fasta -out 927_genome -dbtype nucl\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Long sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "#!/bin/bash\n",
    "\n",
    "#$ -cwd\n",
    "\n",
    "tblastn -query pep_ref_long.fa -db 927_genome -out results_pep_ref_long.csv -outfmt '10 qseqid qseq sseqid pident qlen qstart qend sstart send frames positive mismatch gaps evalue bitscore'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Short sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "#!/bin/bash\n",
    "\n",
    "#$ -cwd\n",
    "\n",
    "tblastn -query pep_ref_short.fa -db 927_genome -out results_pep_ref_short.csv -evalue 100 -word_size 2 -gapopen 9 -gapextend 1 -matrix PAM30 -threshold 16 -comp_based_stats 0 -window_size 15 -outfmt '10 qseqid qseq sseqid pident qlen qstart qend sstart send frames positive mismatch gaps evalue bitscore'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Handling BLAST output"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Re-format command line tBLASTn output to include headers for columns, saving in new files tblastn_pep_ref_short.csv and tblastn_pep_ref_long.csv\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Select only matches with 100% ID (again storing in new files so as to maintain original files in case needed in future)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "#defining function\n",
    "def select_best(infile, outfile):\n",
    "    for line in infile:\n",
    "        ID = (line.split(',')).index('%ID')\n",
    "        outfile.write(line)\n",
    "        break\n",
    "\n",
    "    for line in infile:\n",
    "        columns=line.split(',')\n",
    "        if float(columns[ID]) == 100:\n",
    "            outfile.write(line)\n",
    "    return None\n",
    "\n",
    "#long peptides\n",
    "results = open('tblastn_pep_ref_long.csv', 'r')\n",
    "best_matches = open('pep_ref_long_best.csv', 'w')\n",
    "select_best(results,best_matches)\n",
    "               \n",
    "results.close()\n",
    "best_matches.close()\n",
    "\n",
    "#short peptides - the above function was not used as another condition was included: \n",
    "#that the match was covering the entire query length, \n",
    "#as the parameters used for short sequences can often provide incomplete matches\n",
    "\n",
    "results = open('tblastn_pep_ref_short.csv', 'r')\n",
    "best_matches = open('pep_ref_short_best.csv', 'w')\n",
    "\n",
    "for line in results:\n",
    "        end = (line.split(',')).index('q_end')\n",
    "        start = (line.split(',')).index('q_start')\n",
    "        length= (line.split(',')).index('query_length')\n",
    "        ID = (line.split(',')).index('%ID')\n",
    "        best_matches.write(line)\n",
    "        break\n",
    "\n",
    "for line in results:\n",
    "        columns=line.split(',')\n",
    "        if float(columns[ID]) == 100 and ((float(columns[length])-1))<= (float(columns[end])-float(columns[start])):\n",
    "            best_matches.write(line)\n",
    "\n",
    "results.close()\n",
    "best_matches.close()      \n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Retrieve only coordinates that map to chromosome assemblies (look for and ignore any contigs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "#defining function\n",
    "\n",
    "def select_chrom(infile,outfile):\n",
    "    for line in infile:\n",
    "        name = (line.split(',')).index('subject')\n",
    "        outfile.write(line)\n",
    "        break\n",
    "\n",
    "    for line in infile:\n",
    "        columns = line.split(',')\n",
    "        subject_details = (columns[name]).split('_')\n",
    "        if '11L3' in subject_details or 'bin' in subject_details or 'tryp' in subject_details:    \n",
    "            continue                                 #'11L3', 'bin' and 'tryp' are all key words used in contig names\n",
    "        else:\n",
    "            outfile.write(line)\n",
    "    return None\n",
    "\n",
    "\n",
    "\n",
    "#long peptides\n",
    "\n",
    "results = open('pep_ref_long_best.csv', 'r')\n",
    "chrom = open('pep_ref_long_chromosome.csv', 'w')\n",
    "\n",
    "select_chrom(results, chrom)\n",
    "        \n",
    "results.close()\n",
    "chrom.close()\n",
    "\n",
    "\n",
    "#short peptides\n",
    "\n",
    "results_short = open('pep_ref_short_best.csv', 'r')\n",
    "chrom_short = open('pep_ref_short_chromosome.csv', 'w')\n",
    "\n",
    "select_chrom(results_short,chrom_short)\n",
    "        \n",
    "results_short.close()\n",
    "chrom_short.close()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Comparison of orginal (MaxQuant) and new (tBLASTn) results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "#retreive the ID numbers of the peptides that were analysed\n",
    "\n",
    "pep_dict_IDs = []       \n",
    "for entry in pep_dict.values():     #instead of opening original file again, unique peptides dictionary can be used..\n",
    "    end_ID = entry.index('_')       #...as these peptides were actually analysed after filtering out non-unique\n",
    "    pep_dict_IDs.append(entry[4:end_ID])  \n",
    "\n",
    "    \n",
    "    \n",
    "#obtain the original coordinates (generated by MaxQuant) of the peptides analysed, \n",
    "#storing in dictionary to ensure only unique peptides again and keep ID numbers and coordinates properly associated\n",
    "\n",
    "old_results = open('predicted_new_pep.csv', 'r')\n",
    "\n",
    "results_sorted = {}\n",
    "\n",
    "for line_number, line in enumerate(old_results):      \n",
    "    if line_number >0:\n",
    "        data = line.strip().split(',')\n",
    "        coord = '{}:{}..{}'.format(data[1],data[12],data[13])\n",
    "        pep_details = (data[14]).split('-')  \n",
    "        peptide = pep_details[5]\n",
    "        ID = data[0]\n",
    "        gene_name= data[4]\n",
    "        if ID in pep_dict_IDs:\n",
    "            results_sorted[data[0]] = [gene_name, peptide, coord]\n",
    "        else:\n",
    "            continue\n",
    "               \n",
    "old_results.close()\n",
    "\n",
    "\n",
    "#Combining original and new results (for both long and short peptides) by peptide ID,\n",
    "#saving in new 'results_by_ID_compared.csv' file by first creating results_sorted dictionary \n",
    "               \n",
    "def combine_by_ID(infile):\n",
    "    \n",
    "    for line_number, line in enumerate(infile):\n",
    "        if line_number >0: \n",
    "            data = line.strip().split(',')\n",
    "            ID = ((data[0].split('_'))[0]).strip(' pep')\n",
    "            if not ID in results_sorted :\n",
    "                results_sorted[ID] = []\n",
    "            if data[7]<data[8]:\n",
    "                results_sorted[ID].append('{}:{}..{}'.format(data[2],data[7],data[8]))\n",
    "            elif data[8]<data[7]:\n",
    "                results_sorted[ID].append('{}:{}..{}'.format(data[2],data[8],data[7]))\n",
    "    return results_sorted\n",
    "\n",
    "          \n",
    "results_long = open('tblastn_pep_ref_long.csv', 'r')\n",
    "results_short = open('pep_ref_short_best.csv', 'r')\n",
    "\n",
    "combine_by_ID(results_long)\n",
    "combine_by_ID(results_short)\n",
    "\n",
    "results_long.close()\n",
    "results_short.close()\n",
    "\n",
    "#writing out to file\n",
    "#checking if theres a match between original and new coordinates\n",
    "\n",
    "results_by_ID = open('results_by_ID_compared.csv', 'w')\n",
    "\n",
    "results_by_ID.write(\"ID,gene,peptide,original_coordinates,matching,blast_outputs \\n\")\n",
    "\n",
    "for entry in results_sorted:\n",
    "    ID = entry\n",
    "    gene_name = str(results_sorted[entry][0]).strip('[')\n",
    "    peptide = str(results_sorted[entry][1])\n",
    "    orig_coord = results_sorted[entry][2]\n",
    "    blast = results_sorted[entry][3::]\n",
    "    if orig_coord in blast:\n",
    "        matching = 'MATCH'\n",
    "    else:\n",
    "        matching = 'DIFFERENT'\n",
    "    results_by_ID.write('{},{},{},{},{},\"{}\"\\n'.format(ID,gene_name,peptide,orig_coord,matching,blast)) \n",
    "    \n",
    "\n",
    "results_by_ID.close()\n",
    "        \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Reveiwing Mass Spec data (MaxQuant)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We wish to review the data produced by MaxQuant to manually assess if the evidence is strong enough for the peptide identities assigned by the program"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*Original mass spec evidence files were too large to be added to this Git repository. Instead we have provided the output from this following sub-section 2.1 (ms_data.csv) is provided as a representation of the relevant data only*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Retreiving relevant Information for new protein coding genes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Retrive relevant information (score, delta score, PEP, raw file) for our new coding genes from files of MS evidence.\n",
    "Additionally, annotate if the peptides are observed in Blood Stream Form (BSF) or Procyclic Form (PCF) parasite lifecycle stages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/homes/emarriott/.conda/envs/conda_env_lm/lib/python3.8/site-packages/IPython/core/interactiveshell.py:3145: DtypeWarning: Columns (70,78) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n",
      "/homes/emarriott/.conda/envs/conda_env_lm/lib/python3.8/site-packages/IPython/core/interactiveshell.py:3145: DtypeWarning: Columns (69,70,72) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Length</th>\n",
       "      <th>K Count</th>\n",
       "      <th>R Count</th>\n",
       "      <th>Modifications</th>\n",
       "      <th>Modified sequence</th>\n",
       "      <th>Oxidation (M) Probabilities</th>\n",
       "      <th>Oxidation (M) Score Diffs</th>\n",
       "      <th>Acetyl (Protein N-term)</th>\n",
       "      <th>Oxidation (M)</th>\n",
       "      <th>...</th>\n",
       "      <th>Reverse</th>\n",
       "      <th>Potential contaminant</th>\n",
       "      <th>id</th>\n",
       "      <th>Protein group IDs</th>\n",
       "      <th>Peptide ID</th>\n",
       "      <th>Mod. peptide ID</th>\n",
       "      <th>MS/MS IDs</th>\n",
       "      <th>Best MS/MS</th>\n",
       "      <th>Oxidation (M) site IDs</th>\n",
       "      <th>Lifecycle Stage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAAAAAAAEVESGIAGVEETLR</td>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>Unmodified</td>\n",
       "      <td>_AAAAAAAAEVESGIAGVEETLR_</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1609</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BSF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAAAAAAAEVESGIAGVEETLR</td>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>Unmodified</td>\n",
       "      <td>_AAAAAAAAEVESGIAGVEETLR_</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1609</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BSF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAAAAAAAEVESGIAGVEETLR</td>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>Unmodified</td>\n",
       "      <td>_AAAAAAAAEVESGIAGVEETLR_</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>1609</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BSF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAAAAAAAEVESGIAGVEETLR</td>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>Unmodified</td>\n",
       "      <td>_AAAAAAAAEVESGIAGVEETLR_</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "      <td>1609</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>3.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BSF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAAAAAAAEVESGIAGVEETLR</td>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>Unmodified</td>\n",
       "      <td>_AAAAAAAAEVESGIAGVEETLR_</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4</td>\n",
       "      <td>1609</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>4.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BSF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1266194</th>\n",
       "      <td>YYYGTHVPANNPTWR</td>\n",
       "      <td>15</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>Unmodified</td>\n",
       "      <td>_YYYGTHVPANNPTWR_</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1266194</td>\n",
       "      <td>727</td>\n",
       "      <td>60413</td>\n",
       "      <td>64740</td>\n",
       "      <td>1688917</td>\n",
       "      <td>1688917.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PCF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1266195</th>\n",
       "      <td>YYYGTHVPANNPTWR</td>\n",
       "      <td>15</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>Unmodified</td>\n",
       "      <td>_YYYGTHVPANNPTWR_</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1266195</td>\n",
       "      <td>727</td>\n",
       "      <td>60413</td>\n",
       "      <td>64740</td>\n",
       "      <td>1688918</td>\n",
       "      <td>1688918.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PCF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1266196</th>\n",
       "      <td>YYYGTHVPANNPTWR</td>\n",
       "      <td>15</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>Unmodified</td>\n",
       "      <td>_YYYGTHVPANNPTWR_</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1266196</td>\n",
       "      <td>727</td>\n",
       "      <td>60413</td>\n",
       "      <td>64740</td>\n",
       "      <td>1688919</td>\n",
       "      <td>1688919.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PCF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1266197</th>\n",
       "      <td>YYYGTHVPANNPTWR</td>\n",
       "      <td>15</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>Unmodified</td>\n",
       "      <td>_YYYGTHVPANNPTWR_</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1266197</td>\n",
       "      <td>727</td>\n",
       "      <td>60413</td>\n",
       "      <td>64740</td>\n",
       "      <td>1688920</td>\n",
       "      <td>1688920.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PCF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1266198</th>\n",
       "      <td>YYYGTHVPANNPTWR</td>\n",
       "      <td>15</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>Unmodified</td>\n",
       "      <td>_YYYGTHVPANNPTWR_</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1266198</td>\n",
       "      <td>727</td>\n",
       "      <td>60413</td>\n",
       "      <td>64740</td>\n",
       "      <td>1688921;1688922</td>\n",
       "      <td>1688922.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PCF</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2888055 rows Ã— 80 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       Sequence  Length  K Count  R Count Modifications  \\\n",
       "0        AAAAAAAAEVESGIAGVEETLR      22        0        1    Unmodified   \n",
       "1        AAAAAAAAEVESGIAGVEETLR      22        0        1    Unmodified   \n",
       "2        AAAAAAAAEVESGIAGVEETLR      22        0        1    Unmodified   \n",
       "3        AAAAAAAAEVESGIAGVEETLR      22        0        1    Unmodified   \n",
       "4        AAAAAAAAEVESGIAGVEETLR      22        0        1    Unmodified   \n",
       "...                         ...     ...      ...      ...           ...   \n",
       "1266194         YYYGTHVPANNPTWR      15        0        1    Unmodified   \n",
       "1266195         YYYGTHVPANNPTWR      15        0        1    Unmodified   \n",
       "1266196         YYYGTHVPANNPTWR      15        0        1    Unmodified   \n",
       "1266197         YYYGTHVPANNPTWR      15        0        1    Unmodified   \n",
       "1266198         YYYGTHVPANNPTWR      15        0        1    Unmodified   \n",
       "\n",
       "                Modified sequence Oxidation (M) Probabilities  \\\n",
       "0        _AAAAAAAAEVESGIAGVEETLR_                         NaN   \n",
       "1        _AAAAAAAAEVESGIAGVEETLR_                         NaN   \n",
       "2        _AAAAAAAAEVESGIAGVEETLR_                         NaN   \n",
       "3        _AAAAAAAAEVESGIAGVEETLR_                         NaN   \n",
       "4        _AAAAAAAAEVESGIAGVEETLR_                         NaN   \n",
       "...                           ...                         ...   \n",
       "1266194         _YYYGTHVPANNPTWR_                         NaN   \n",
       "1266195         _YYYGTHVPANNPTWR_                         NaN   \n",
       "1266196         _YYYGTHVPANNPTWR_                         NaN   \n",
       "1266197         _YYYGTHVPANNPTWR_                         NaN   \n",
       "1266198         _YYYGTHVPANNPTWR_                         NaN   \n",
       "\n",
       "        Oxidation (M) Score Diffs  Acetyl (Protein N-term)  Oxidation (M)  \\\n",
       "0                             NaN                        0              0   \n",
       "1                             NaN                        0              0   \n",
       "2                             NaN                        0              0   \n",
       "3                             NaN                        0              0   \n",
       "4                             NaN                        0              0   \n",
       "...                           ...                      ...            ...   \n",
       "1266194                       NaN                        0              0   \n",
       "1266195                       NaN                        0              0   \n",
       "1266196                       NaN                        0              0   \n",
       "1266197                       NaN                        0              0   \n",
       "1266198                       NaN                        0              0   \n",
       "\n",
       "         ...  Reverse Potential contaminant       id Protein group IDs  \\\n",
       "0        ...      NaN                   NaN        0              1609   \n",
       "1        ...      NaN                   NaN        1              1609   \n",
       "2        ...      NaN                   NaN        2              1609   \n",
       "3        ...      NaN                   NaN        3              1609   \n",
       "4        ...      NaN                   NaN        4              1609   \n",
       "...      ...      ...                   ...      ...               ...   \n",
       "1266194  ...      NaN                   NaN  1266194               727   \n",
       "1266195  ...      NaN                   NaN  1266195               727   \n",
       "1266196  ...      NaN                   NaN  1266196               727   \n",
       "1266197  ...      NaN                   NaN  1266197               727   \n",
       "1266198  ...      NaN                   NaN  1266198               727   \n",
       "\n",
       "        Peptide ID  Mod. peptide ID        MS/MS IDs  Best MS/MS  \\\n",
       "0                0                0                0         0.0   \n",
       "1                0                0                1         1.0   \n",
       "2                0                0                2         2.0   \n",
       "3                0                0                3         3.0   \n",
       "4                0                0                4         4.0   \n",
       "...            ...              ...              ...         ...   \n",
       "1266194      60413            64740          1688917   1688917.0   \n",
       "1266195      60413            64740          1688918   1688918.0   \n",
       "1266196      60413            64740          1688919   1688919.0   \n",
       "1266197      60413            64740          1688920   1688920.0   \n",
       "1266198      60413            64740  1688921;1688922   1688922.0   \n",
       "\n",
       "        Oxidation (M) site IDs  Lifecycle Stage  \n",
       "0                          NaN              BSF  \n",
       "1                          NaN              BSF  \n",
       "2                          NaN              BSF  \n",
       "3                          NaN              BSF  \n",
       "4                          NaN              BSF  \n",
       "...                        ...              ...  \n",
       "1266194                    NaN              PCF  \n",
       "1266195                    NaN              PCF  \n",
       "1266196                    NaN              PCF  \n",
       "1266197                    NaN              PCF  \n",
       "1266198                    NaN              PCF  \n",
       "\n",
       "[2888055 rows x 80 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#read in the proteomic evidence data for BSF and add a 'lifecyle stage' column, set this to 'BSF' \n",
    "df_bsf = pd.read_csv('evidence_bsf.zip',sep='\\t')\n",
    "df_bsf['Lifecycle Stage'] = 'BSF'\n",
    "\n",
    "#as above for PCF\n",
    "df_pcf = pd.read_csv('evidence_pcf.zip',sep='\\t')\n",
    "df_pcf['Lifecycle Stage'] = 'PCF'\n",
    "\n",
    "#combine as one dataframe\n",
    "df_all = df_bsf.append(df_pcf)\n",
    "df_all"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>peptides</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>KS17gene_1898a</td>\n",
       "      <td>[GTMPQEPTR, GTMPQEPTRGDR, APSGFPVLQGNSSKPTN]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>KS17gene_2338a</td>\n",
       "      <td>[NDNYNKPVPGAEGQNDGLHFPQR]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>KS17gene_265a</td>\n",
       "      <td>[FSGLVPDK, FSGLVPDKLK, KATSSLTDQLTK, ATSSLTDQLTK]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>KS17gene_6299a</td>\n",
       "      <td>[MEGLGLETQMR]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>KS17gene_6998a</td>\n",
       "      <td>[GPLPDDLPDTSRETDTLDER, GPLPDDLPDTSR]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>KS17gene_7003a</td>\n",
       "      <td>[SLYAGYFDGAAAQQK, SLDADQASVDADIFR, ALTYCILR, E...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>KS17gene_8518a</td>\n",
       "      <td>[TREPPADGFVDGASQFEGVSVK, VHTESDVYIPAAAFR]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>MSTRG.94</td>\n",
       "      <td>[LAGVELQGINLSPPQR, LCQLELLISR, GVVDAFTTLLR]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>TRY.375</td>\n",
       "      <td>[TSFLVEGR, TGTATDNATVALNCNPLE]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             gene                                           peptides\n",
       "0  KS17gene_1898a       [GTMPQEPTR, GTMPQEPTRGDR, APSGFPVLQGNSSKPTN]\n",
       "1  KS17gene_2338a                          [NDNYNKPVPGAEGQNDGLHFPQR]\n",
       "2   KS17gene_265a  [FSGLVPDK, FSGLVPDKLK, KATSSLTDQLTK, ATSSLTDQLTK]\n",
       "3  KS17gene_6299a                                      [MEGLGLETQMR]\n",
       "4  KS17gene_6998a               [GPLPDDLPDTSRETDTLDER, GPLPDDLPDTSR]\n",
       "5  KS17gene_7003a  [SLYAGYFDGAAAQQK, SLDADQASVDADIFR, ALTYCILR, E...\n",
       "6  KS17gene_8518a          [TREPPADGFVDGASQFEGVSVK, VHTESDVYIPAAAFR]\n",
       "7        MSTRG.94        [LAGVELQGINLSPPQR, LCQLELLISR, GVVDAFTTLLR]\n",
       "8         TRY.375                     [TSFLVEGR, TGTATDNATVALNCNPLE]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#from previous blast results, create dataframe of gene names and peptide sequences for genes of interest only\n",
    "\n",
    "genes_of_interest = ['TRY.375','MSTRG.94','KS17gene_8518a','KS17gene_7003a', 'KS17gene_6998a', \n",
    "                                     'KS17gene_6299a', 'KS17gene_265a', 'KS17gene_2338a', \n",
    "                                      'KS17gene_1898a']\n",
    "\n",
    "df_genes = pd.read_csv('results_by_ID_compared.csv', index_col=0)\n",
    "\n",
    "df_genes = df_genes[df_genes['gene'].isin(genes_of_interest)][['gene','peptide']]\n",
    "\n",
    "df_genes = df_genes.groupby('gene')['peptide'].apply(list).reset_index(name='peptides')\n",
    "\n",
    "df_genes\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Score</th>\n",
       "      <th>Delta score</th>\n",
       "      <th>PEP</th>\n",
       "      <th>Raw file</th>\n",
       "      <th>Lifecycle Stage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Gene, Sequence, Score, Delta score, PEP, Raw file, Lifecycle Stage]\n",
       "Index: []"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#create empty output data frame to append relevant information to\n",
    "\n",
    "df_output = pd.DataFrame(columns= ['Gene','Sequence','Score','Delta score','PEP','Raw file', 'Lifecycle Stage'])\n",
    "\n",
    "df_output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Score</th>\n",
       "      <th>Delta score</th>\n",
       "      <th>PEP</th>\n",
       "      <th>Raw file</th>\n",
       "      <th>Lifecycle Stage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>452794</th>\n",
       "      <td>KS17gene_1898a</td>\n",
       "      <td>GTMPQEPTRGDR</td>\n",
       "      <td>101.930</td>\n",
       "      <td>92.974</td>\n",
       "      <td>4.568600e-06</td>\n",
       "      <td>PT5270-69</td>\n",
       "      <td>PCF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98741</th>\n",
       "      <td>KS17gene_1898a</td>\n",
       "      <td>APSGFPVLQGNSSKPTN</td>\n",
       "      <td>107.650</td>\n",
       "      <td>82.172</td>\n",
       "      <td>3.861100e-05</td>\n",
       "      <td>PT5268-53</td>\n",
       "      <td>PCF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>452769</th>\n",
       "      <td>KS17gene_1898a</td>\n",
       "      <td>GTMPQEPTR</td>\n",
       "      <td>81.296</td>\n",
       "      <td>63.261</td>\n",
       "      <td>4.041600e-03</td>\n",
       "      <td>PT5270-69</td>\n",
       "      <td>PCF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1006916</th>\n",
       "      <td>KS17gene_2338a</td>\n",
       "      <td>NDNYNKPVPGAEGQNDGLHFPQR</td>\n",
       "      <td>184.560</td>\n",
       "      <td>146.350</td>\n",
       "      <td>1.740900e-19</td>\n",
       "      <td>LG-30min-C7</td>\n",
       "      <td>BSF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>721869</th>\n",
       "      <td>KS17gene_265a</td>\n",
       "      <td>KATSSLTDQLTK</td>\n",
       "      <td>242.450</td>\n",
       "      <td>158.950</td>\n",
       "      <td>4.479100e-53</td>\n",
       "      <td>LG_2hr_B8</td>\n",
       "      <td>BSF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128707</th>\n",
       "      <td>KS17gene_265a</td>\n",
       "      <td>ATSSLTDQLTK</td>\n",
       "      <td>175.650</td>\n",
       "      <td>64.921</td>\n",
       "      <td>4.197500e-06</td>\n",
       "      <td>LG_4hr_A8</td>\n",
       "      <td>BSF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>448985</th>\n",
       "      <td>KS17gene_265a</td>\n",
       "      <td>FSGLVPDKLK</td>\n",
       "      <td>122.330</td>\n",
       "      <td>122.330</td>\n",
       "      <td>2.055100e-03</td>\n",
       "      <td>LG_8hr_C6</td>\n",
       "      <td>BSF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>448979</th>\n",
       "      <td>KS17gene_265a</td>\n",
       "      <td>FSGLVPDK</td>\n",
       "      <td>128.610</td>\n",
       "      <td>75.924</td>\n",
       "      <td>5.387000e-03</td>\n",
       "      <td>LG-30min-C2</td>\n",
       "      <td>BSF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>745832</th>\n",
       "      <td>KS17gene_6299a</td>\n",
       "      <td>MEGLGLETQMR</td>\n",
       "      <td>105.980</td>\n",
       "      <td>88.166</td>\n",
       "      <td>4.959100e-04</td>\n",
       "      <td>PT5268-59</td>\n",
       "      <td>PCF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>434506</th>\n",
       "      <td>KS17gene_6998a</td>\n",
       "      <td>GPLPDDLPDTSR</td>\n",
       "      <td>74.162</td>\n",
       "      <td>62.115</td>\n",
       "      <td>2.517500e-03</td>\n",
       "      <td>PT5270-59</td>\n",
       "      <td>PCF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>434508</th>\n",
       "      <td>KS17gene_6998a</td>\n",
       "      <td>GPLPDDLPDTSRETDTLDER</td>\n",
       "      <td>40.085</td>\n",
       "      <td>31.376</td>\n",
       "      <td>8.760400e-03</td>\n",
       "      <td>PT5270-7</td>\n",
       "      <td>PCF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>913758</th>\n",
       "      <td>KS17gene_7003a</td>\n",
       "      <td>SLDADQASVDADIFR</td>\n",
       "      <td>193.230</td>\n",
       "      <td>149.030</td>\n",
       "      <td>1.151500e-72</td>\n",
       "      <td>PT5269-28</td>\n",
       "      <td>PCF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1253866</th>\n",
       "      <td>KS17gene_7003a</td>\n",
       "      <td>SLYAGYFDGAAAQQK</td>\n",
       "      <td>211.460</td>\n",
       "      <td>148.830</td>\n",
       "      <td>4.820500e-29</td>\n",
       "      <td>LG-30min-C2</td>\n",
       "      <td>BSF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>280567</th>\n",
       "      <td>KS17gene_7003a</td>\n",
       "      <td>ELPSHAPYSASLPLK</td>\n",
       "      <td>44.196</td>\n",
       "      <td>31.776</td>\n",
       "      <td>1.637400e-02</td>\n",
       "      <td>PT5269-24</td>\n",
       "      <td>PCF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82343</th>\n",
       "      <td>KS17gene_7003a</td>\n",
       "      <td>ALTYCILR</td>\n",
       "      <td>66.993</td>\n",
       "      <td>28.860</td>\n",
       "      <td>4.002900e-02</td>\n",
       "      <td>PT5268-5</td>\n",
       "      <td>PCF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1381035</th>\n",
       "      <td>KS17gene_8518a</td>\n",
       "      <td>TREPPADGFVDGASQFEGVSVK</td>\n",
       "      <td>172.810</td>\n",
       "      <td>135.520</td>\n",
       "      <td>1.723900e-12</td>\n",
       "      <td>LG-1hr-A7</td>\n",
       "      <td>BSF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1128903</th>\n",
       "      <td>KS17gene_8518a</td>\n",
       "      <td>VHTESDVYIPAAAFR</td>\n",
       "      <td>104.170</td>\n",
       "      <td>83.222</td>\n",
       "      <td>1.406200e-05</td>\n",
       "      <td>PT5269-32</td>\n",
       "      <td>PCF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>781215</th>\n",
       "      <td>MSTRG.94</td>\n",
       "      <td>LAGVELQGINLSPPQR</td>\n",
       "      <td>240.370</td>\n",
       "      <td>240.370</td>\n",
       "      <td>7.851100e-46</td>\n",
       "      <td>LG_8hr_A7</td>\n",
       "      <td>BSF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>576920</th>\n",
       "      <td>MSTRG.94</td>\n",
       "      <td>GVVDAFTTLLR</td>\n",
       "      <td>196.880</td>\n",
       "      <td>105.020</td>\n",
       "      <td>4.652600e-12</td>\n",
       "      <td>LG-1hr-B10</td>\n",
       "      <td>BSF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>792606</th>\n",
       "      <td>MSTRG.94</td>\n",
       "      <td>LCQLELLISR</td>\n",
       "      <td>144.770</td>\n",
       "      <td>87.179</td>\n",
       "      <td>1.928800e-04</td>\n",
       "      <td>LG-1hr-B8</td>\n",
       "      <td>BSF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1011449</th>\n",
       "      <td>TRY.375</td>\n",
       "      <td>TGTATDNATVALNCNPLE</td>\n",
       "      <td>123.280</td>\n",
       "      <td>90.669</td>\n",
       "      <td>2.184000e-13</td>\n",
       "      <td>PT5268-62</td>\n",
       "      <td>PCF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1050125</th>\n",
       "      <td>TRY.375</td>\n",
       "      <td>TSFLVEGR</td>\n",
       "      <td>101.410</td>\n",
       "      <td>45.290</td>\n",
       "      <td>5.108400e-03</td>\n",
       "      <td>PT5269-26</td>\n",
       "      <td>PCF</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   Gene                 Sequence    Score  Delta score  \\\n",
       "452794   KS17gene_1898a             GTMPQEPTRGDR  101.930       92.974   \n",
       "98741    KS17gene_1898a        APSGFPVLQGNSSKPTN  107.650       82.172   \n",
       "452769   KS17gene_1898a                GTMPQEPTR   81.296       63.261   \n",
       "1006916  KS17gene_2338a  NDNYNKPVPGAEGQNDGLHFPQR  184.560      146.350   \n",
       "721869    KS17gene_265a             KATSSLTDQLTK  242.450      158.950   \n",
       "128707    KS17gene_265a              ATSSLTDQLTK  175.650       64.921   \n",
       "448985    KS17gene_265a               FSGLVPDKLK  122.330      122.330   \n",
       "448979    KS17gene_265a                 FSGLVPDK  128.610       75.924   \n",
       "745832   KS17gene_6299a              MEGLGLETQMR  105.980       88.166   \n",
       "434506   KS17gene_6998a             GPLPDDLPDTSR   74.162       62.115   \n",
       "434508   KS17gene_6998a     GPLPDDLPDTSRETDTLDER   40.085       31.376   \n",
       "913758   KS17gene_7003a          SLDADQASVDADIFR  193.230      149.030   \n",
       "1253866  KS17gene_7003a          SLYAGYFDGAAAQQK  211.460      148.830   \n",
       "280567   KS17gene_7003a          ELPSHAPYSASLPLK   44.196       31.776   \n",
       "82343    KS17gene_7003a                 ALTYCILR   66.993       28.860   \n",
       "1381035  KS17gene_8518a   TREPPADGFVDGASQFEGVSVK  172.810      135.520   \n",
       "1128903  KS17gene_8518a          VHTESDVYIPAAAFR  104.170       83.222   \n",
       "781215         MSTRG.94         LAGVELQGINLSPPQR  240.370      240.370   \n",
       "576920         MSTRG.94              GVVDAFTTLLR  196.880      105.020   \n",
       "792606         MSTRG.94               LCQLELLISR  144.770       87.179   \n",
       "1011449         TRY.375       TGTATDNATVALNCNPLE  123.280       90.669   \n",
       "1050125         TRY.375                 TSFLVEGR  101.410       45.290   \n",
       "\n",
       "                  PEP     Raw file Lifecycle Stage  \n",
       "452794   4.568600e-06    PT5270-69             PCF  \n",
       "98741    3.861100e-05    PT5268-53             PCF  \n",
       "452769   4.041600e-03    PT5270-69             PCF  \n",
       "1006916  1.740900e-19  LG-30min-C7             BSF  \n",
       "721869   4.479100e-53    LG_2hr_B8             BSF  \n",
       "128707   4.197500e-06    LG_4hr_A8             BSF  \n",
       "448985   2.055100e-03    LG_8hr_C6             BSF  \n",
       "448979   5.387000e-03  LG-30min-C2             BSF  \n",
       "745832   4.959100e-04    PT5268-59             PCF  \n",
       "434506   2.517500e-03    PT5270-59             PCF  \n",
       "434508   8.760400e-03     PT5270-7             PCF  \n",
       "913758   1.151500e-72    PT5269-28             PCF  \n",
       "1253866  4.820500e-29  LG-30min-C2             BSF  \n",
       "280567   1.637400e-02    PT5269-24             PCF  \n",
       "82343    4.002900e-02     PT5268-5             PCF  \n",
       "1381035  1.723900e-12    LG-1hr-A7             BSF  \n",
       "1128903  1.406200e-05    PT5269-32             PCF  \n",
       "781215   7.851100e-46    LG_8hr_A7             BSF  \n",
       "576920   4.652600e-12   LG-1hr-B10             BSF  \n",
       "792606   1.928800e-04    LG-1hr-B8             BSF  \n",
       "1011449  2.184000e-13    PT5268-62             PCF  \n",
       "1050125  5.108400e-03    PT5269-26             PCF  "
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#for genes of interest (df_genes), retrieve relevent information from evidence dataframe (df_all)\n",
    "#df_all can be searched by peptide sequence, \n",
    "#so retrieve information for the peptides in the list under each gene in df_genes\n",
    "\n",
    "for index, row in df_genes.iterrows():\n",
    "    gene_name = row['gene']\n",
    "    df_all['Gene'] = gene_name\n",
    "    df_output = df_output.append(df_all[df_all['Sequence'].isin(row['peptides'])][\n",
    "        ['Gene','Sequence','Score','Delta score','PEP','Raw file', 'Lifecycle Stage']].sort_values('PEP').drop_duplicates('Sequence'))\n",
    "         #save in dataframe only the information and columns that are relevant to our analysis\n",
    "         #sort results by quality (low to high PEP) and remove non-unique peptides\n",
    "\n",
    "\n",
    "df_output.to_csv('ms_data.csv')  #save as csv file\n",
    "df_output"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {
    "height": "143px",
    "width": "458px"
   },
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": true,
   "toc_position": {
    "height": "calc(100% - 180px)",
    "left": "10px",
    "top": "150px",
    "width": "297px"
   },
   "toc_section_display": true,
   "toc_window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
